### PR TITLE
Add option for re-registration when ephemeral Consul is restarted

### PR DESCRIPTION
**Pull request by making**
Fixes gh-197 .

I haven't updated the doc as TTL is not yet documented.

### CODE REVIEW

**Comment by making** [comment]
I found the doc about TTL is included in #690

**Comment by pcornelissen** [comment]
I have not checked your PR, but I just wanted to document my idea somewhere:
A good indicator if the service needs to register in consul (again) is that when it fetches the instances list from consul and itself is not included in that list.
The neat thing is that this criteria would remove the necessity to register programatically on startup, because the service can just connect to consul and start fetching the instance list and then (obviously) it's not present yet and the criteria above would trigger a registration.

Just my 2C, I'm not sure if that is feasible or the best way

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/HeartbeatProperties.java [hunk 1]
```diff
@@ -51,6 +51,8 @@
 	@DecimalMax("0.9")
 	private double intervalRatio = 2.0 / 3.0;
 
+	private boolean reRegisterServiceOnFailure = false;
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/HeartbeatProperties.java:54]: `reregisterServiceOnFailure`

File: docs/src/main/asciidoc/_configprops.adoc [hunk 2]
```diff
@@ -33,7 +33,8 @@
 |spring.cloud.consul.discovery.health-check-url |  | Custom health check url to override default.
 |spring.cloud.consul.discovery.heartbeat.enabled | `false` | 
 |spring.cloud.consul.discovery.heartbeat.interval-ratio |  | 
-|spring.cloud.consul.discovery.heartbeat.ttl | `30s` | 
+|spring.cloud.consul.discovery.heartbeat.ttl | `30s` |
+|spring.cloud.consul.discovery.heartbeat.re-register-service-on-failure | `false` | Enables service re-registration when the heartbeat fails.
```
Comment (spencergibb) [inline docs/src/main/asciidoc/_configprops.adoc:37]: `reregister`

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ReRegistrationPredicate.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+package org.springframework.cloud.consul.discovery;
+
+import com.ecwid.consul.v1.OperationException;
+
+/**
+ * Predicate on whether to re-register service.
+ *
+ * @author Toshiaki Maki
+ */
+public interface ReRegistrationPredicate {
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ReRegistrationPredicate.java:26]: Why new interface? Just use `Predicate<OperationException`.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java [hunk 4]
```diff
@@ -77,22 +87,46 @@ public void remove(String instanceId) {
...
+			}
+			catch (OperationException e) {
+				if (this.ttlScheduler.heartbeatProperties.isReRegisterServiceOnFailure()
+						&& this.ttlScheduler.reRegistrationPredicate.isEligible(e)) {
+					log.warn(e.getMessage());
+					if (this.ttlScheduler.registeredService != null) {
+						if (log.isInfoEnabled()) {
+							log.info("Re-register " + this.ttlScheduler.registeredService);
+						}
+						this.ttlScheduler.client.agentServiceRegister(this.ttlScheduler.registeredService,
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:120]: `this.client`?

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/serviceregistry/ConsulServiceRegistry.java [hunk 5]
```diff
@@ -68,7 +68,7 @@ public void register(ConsulRegistration reg) {
 			NewService service = reg.getService();
 			if (this.heartbeatProperties.isEnabled() && this.ttlScheduler != null && service.getCheck() != null
 					&& service.getCheck().getTtl() != null) {
-				this.ttlScheduler.add(reg.getInstanceId());
+				this.ttlScheduler.add(reg.getService());
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/serviceregistry/ConsulServiceRegistry.java:71]: Why this change?

**Review by spencergibb — CHANGES_REQUESTED** [review]
Looks like a good start.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ReRegistrationPredicate.java (hunk 3 above)
Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ReRegistrationPredicate.java]: I tend to avoid using classes under `java.*` package as a bean type. It can potentially conflict with other libraries. For example, Spring Cloud Function uses `Function`, `Consumer` and `Supplier`. If there is no possibility of a conflict in `Predicate`, I don't mind dropping this class.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java (hunk 4 above)
Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:120]: `client` is not visible in `ConsulHeartbeatTask`

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/serviceregistry/ConsulServiceRegistry.java (hunk 5 above)
Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/serviceregistry/ConsulServiceRegistry.java:71]: `ConsulClient.agentServiceRegister` requires `NewService` instance rather than just a service id.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/HeartbeatProperties.java (hunk 1 above)
Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/HeartbeatProperties.java]: done

File: docs/src/main/asciidoc/_configprops.adoc (hunk 2 above)
Comment (making) [reply docs/src/main/asciidoc/_configprops.adoc]: done

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java [hunk 6]
```diff
@@ -42,27 +43,36 @@
...
+		this.heartbeatProperties = heartbeatProperties;
+		this.discoveryProperties = discoveryProperties;
 		this.client = client;
+		this.reregistrationPredicate = reregistrationPredicate;
 	}
 
-	@Deprecated
 	public void add(final NewService service) {
 		add(service.getId());
+		this.registeredService = service;
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:66]: There's a reason add adds into a map. Any reason why `NewService` doesn't get added to one as well.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java (hunk 4 above)
Comment (spencergibb) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:120]: but isn't that because you changed the class to `static`?

Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:120]: yeah I've change it to static because of the testability.
Would you like me to revert to private and use `ReflectionUtils` in `ConsulHeartbeatTaskTests`?

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java (hunk 6 above)
Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:66]: Do you meath it should have `Map<String, NewService> registeredServices` ?

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java (hunk 4 above)
Comment (spencergibb) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:120]: No

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java (hunk 6 above)
Comment (spencergibb) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java:66]: Yes, I think so. You'll need to update remove as well.

Comment (making) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java]: Updated
